src/operator/scan/duckdb_scan_task.cpp

```diff
@@ -0,0 +1,469 @@
+duckdb_scan_task_local_state::column_builder::column_builder(duckdb::LogicalType t)
+  : type(t), type_size(duckdb::GetTypeIdSize(type.InternalType()))
+{
+  auto& mem_res_mgr = memory::memory_reservation_manager::get_instance();
+  auto mem_space    = mem_res_mgr.get_memory_spaces_for_tier(memory::Tier::HOST)[HOST_SPACE_INDEX];
+  allocator =
+    mem_space->get_default_allocator_as<sirius::memory::fixed_size_host_memory_resource>();
+
+  // Check that the allocator is not nullptr
+  if (allocator == nullptr) {
+    throw std::runtime_error(
+      "Failed to get fixed_size_host_memory_resource allocator for HOST memory space.");
+  }
+}
+
+duckdb_scan_task_local_state::column_builder::~column_builder()
+{
+  // Release reservations. Memory is deallocated automatically.
+  auto& mem_res_mgr = memory::memory_reservation_manager::get_instance();
+  if (data_reservation) { mem_res_mgr.release_reservation(std::move(data_reservation)); }
+  if (mask_reservation) { mem_res_mgr.release_reservation(std::move(mask_reservation)); }
+  if (offset_reservation) { mem_res_mgr.release_reservation(std::move(offset_reservation)); }
+}
+
+void duckdb_scan_task_local_state::column_builder::reserve_memory(size_t estimated_num_rows)
+{
+  assert(estimated_num_rows > 0);
+
+  auto& mem_res_mgr = memory::memory_reservation_manager::get_instance();
+  data_reservation  = mem_res_mgr.request_reservation(res_request, type_size * estimated_num_rows);
+  mask_reservation =
+    mem_res_mgr.request_reservation(res_request, utils::ceil_div_8(estimated_num_rows));
+  if (type.InternalType() == duckdb::PhysicalType::VARCHAR) {
+    offset_reservation =
+      mem_res_mgr.request_reservation(res_request, sizeof(int64_t) * (estimated_num_rows + 1));
+  }
+}
+
+void duckdb_scan_task_local_state::column_builder::allocate_memory()
+{
+  assert(data_reservation);
+  assert(mask_reservation);
+
+  data_blocks_accessor.initialize(sirius::make_unique<multiple_blocks_allocation>(
+    allocator->allocate_multiple_blocks(data_reservation->size)));
+
+  mask_blocks_accessor.initialize(sirius::make_unique<multiple_blocks_allocation>(
+    allocator->allocate_multiple_blocks(mask_reservation->size)));
+
+  if (type.InternalType() == duckdb::PhysicalType::VARCHAR) {
+    assert(offset_reservation);
+
+    offset_blocks_accessor.initialize(sirius::make_unique<multiple_blocks_allocation>(
+      allocator->allocate_multiple_blocks(offset_reservation->size)));
+
+    // Initialize the running prefix sum of offsets to zero
+    offset_blocks_accessor.set_current(0);
+  }
+}
+
+// This method should usually be called only on variable-length data
+bool duckdb_scan_task_local_state::column_builder::sufficient_space_for_column(
+  duckdb::Vector& vec, duckdb::ValidityMask const& validity, size_t num_rows)
+{
+  size_t data_bytes = 0;
+  if (type.InternalType() == duckdb::PhysicalType::VARCHAR) {
+    auto const* str_data = reinterpret_cast<duckdb::string_t const*>(vec.GetData());
```

Comment:
shouldn't we be checking the validity mask before adding the size? If we aren't chekcing that max isn't there an api that can get this size without having to do a full pass over the data? There might not be I am asking